realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -351,6 +414,60 @@ private void clearScheduledAccessTokenRefresh() {
         }
     }
 
+    // Wrapper class for handling the async operations of the underlying SyncSession calling `async_wait_for_download_completion`
+    private static class WaitForServerChangesWrapper {
+
+        private final CountDownLatch waitForChanges = new CountDownLatch(1);
+        private volatile boolean resultReceived = false;
+        private Long errorCode = null;
+        private String errorMessage;
+
+        public void cancelWait() {
+            waitForChanges.countDown();
+        }
+
+        /**
+         * Block until the wait either completes or is terminated for other reasons.
+         */
+        public void waitForServerChanges() {
+            if (!resultReceived) {
+                try {
+                    waitForChanges.await();
+                } catch (InterruptedException e) {
+                    throw new ObjectServerError(ErrorCode.UNKNOWN, e);
```

Comment:
Re-thought about this, this might need to throw a checked `InterruptedException` or a special exception for `RealmAsyncTask` returned by `getInstanceAsync()` for cancelling. I will revisit this part after #4570 merged.